system/mp/mpcomm.cpp

```diff
@@ -2945,46 +2949,99 @@ class CCommunicator: public ICommunicator, public CInterface
+                Sleep(100);
+            }
+        };
+
+        auto threadedVerifyConnectFunc = [&](rank_t rank, std::function<bool (unsigned rank, unsigned timeout)> connectFunc)
+        {
+            // NB: running because took (via wait()) a semaphore slot, restore it at end of scope
+            struct RestoreSlot
+            {
+                Semaphore &sem;
+                RestoreSlot(Semaphore &_sem) : sem(_sem) { }
+                ~RestoreSlot() { sem.signal(); }
+            } restoreSlot(sem);
+
+            unsigned timeoutMs;
+            if (totalTM.timedout(&timeoutMs) || abort)
+                return false;
+            if (perConnectionTimeout && (perConnectionTimeout < timeoutMs))
+                timeoutMs = perConnectionTimeout;
+
+            if (!connectFunc(rank, timeoutMs))
+            {
+                abort = true; // ensure verifyFunc knows before release slot, to prevent other thread being launched
+                return false;
+            }
+            return true;
+        };
+
+        auto verifyFunc = [&](std::function<bool (unsigned rank)> conditionFunc, std::function<bool (unsigned rank, unsigned timeout)> connectFunc)
+        {
+            std::vector<std::future<bool>> threads;
+            for (rank_t rank=0; rank<group->ordinality(); rank++)
+            {
+                if (conditionFunc(rank))
+                {
+                    // check timeout before and after sem.wait
+                    // NB: sem.wait if successful, takes a slot which is restored by the thread when it is done
                     unsigned remaining;
-                    if (tm.timedout(&remaining)) {
-                        return false;
+                    if (totalTM.timedout(&remaining) || !sem.wait(remaining) || totalTM.timedout(&remaining))
+                    {
+                        abort = true;
+                        break;
                     }
-                    if (!channel->verifyConnection(tm,true))
-                        return false;
+                    else if (abort)
+                        break;
+                    threads.push_back(std::async(std::launch::async, threadedVerifyConnectFunc, rank, connectFunc));
                 }
             }
-        }
-        if (!duplex) {
-            ForEachNodeInGroup(rank,*group) {
-                bool doverify = ((rank&1)==(myrank&1))?(myrank<rank):(myrank>rank);
-                if (doverify) {
-                    Owned<CMPChannel> channel = getChannel(rank);
-                    while (!channel->verifyConnection(tm,false)) {
-                        unsigned remaining;
-                        if (tm.timedout(&remaining))
-                            return false;
-                        CriticalUnblock unblock(verifysect);
-                        Sleep(100);
-                    }
-                }
+            bool res = true;
+            for (auto &f: threads)
+            {
+                if (!f.get())
```

Comment:
It sort of seems that the above loop over ranks could be a while and interleave in that loop the check of each thread's completion status and if complete then a get() ?

Reply:
I think its ok as is, but since get() blocks it does seem like we are waiting on them in order.

Reply:
get() on future is effectively a join() and then getting the result from the thread.

The threads will run asynchronously, yes you could be blocked on 1st get() (joining), whilst 2->n had already finished.

But you have to wait on all anyway (have to join all), so there's no value in timing out and joining others instead.
If were waiting blocked (joining) on 1 and others had finished, it would join all others as soon as became unblocked.

I don't think there's a better way of coding it or any gain.
But I could be missing something in what you mean.
@mckellyln  - do you have some pseudo code of what you mean?

Reply:
See my comment above this one.  I think there may be a problem if group->ordinality() > std::thread::hardware_concurrency() ?  And then to change it to a while loop to handle that __suggests__ to me that one might interleave the checking for complete in some threads while still waiting for others.  But yes I agree and stated above, its ok to get() afterwards - as long as we started all we need to.

Reply:
>one might interleave the checking for complete in some threads while still waiting for others

still not sure what point would be. But an example may help me understand benefits.

Reply:
I'll agree it is ok as coded.